lib/Fmt_ast.ml

```diff
@@ -2077,20 +2077,25 @@ and fmt_expression c ?(box = true) ?pro ?epi ?eol ?parens ?(indent_wrap = 0)
         | _ -> Option.is_some (Sugar.list_exp c.cmts e0)
       in
       let opn, cls = if can_skip_parens then (".", "") else (".(", ")") in
-      hvbox 0
-        ( fits_breaks_if parens "" "("
-        $ fits_breaks "" "let "
-        $ Cmts.fmt c popen_loc
-            ( fits_breaks "" (if override then "open! " else "open ")
-            $ fmt_module_statement c ~attributes noop
-                (sub_mod ~ctx popen_expr) )
-        $ fits_breaks opn " in"
-        $ fmt_or_k force_fit_if (fmt "@;<0 2>")
-            (fits_breaks "" ~hint:(1000, 0) "")
-        $ fmt_expression c (sub_exp ~ctx e0)
-        $ fits_breaks cls ""
-        $ fits_breaks_if parens "" ")"
-        $ fmt_atrs )
+      let outer_parens, inner_parens =
+        if has_attr then (parens, true) else (false, parens)
+      in
+      hovbox 0
+        ( fmt_if outer_parens "("
+        $ hvbox 0
+            ( fits_breaks_if inner_parens "" "("
+            $ fits_breaks "" "let "
+            $ Cmts.fmt c popen_loc
+                ( fits_breaks "" (if override then "open! " else "open ")
+                $ fmt_module_statement c ~attributes noop
+                    (sub_mod ~ctx popen_expr) )
+            $ fits_breaks opn " in"
+            $ fmt_or_k force_fit_if (fmt "@;<0 2>")
+                (fits_breaks "" ~hint:(1000, 0) "")
+            $ fmt_expression c (sub_exp ~ctx e0)
+            $ fits_breaks cls ""
+            $ fits_breaks_if inner_parens "" ")" )
+        $ fmt_atrs $ fmt_if outer_parens ")" )
```

Comment:
the `outer_parens` 2 parts could be replaced by a `wrap_if`, that would be easier to read I think